tests/test_before_after_filesystem_snapshot.py

```diff
@@ -0,0 +1,113 @@
+      'three.txt': '1111222233334444',
+      'bar/bat/four.tgz': '6677889900112233'
+    }
+
+    snapshot = before_after_filesystem_snapshot.snapshot(self.before, after)
+    self.assertEqual(snapshot, (['bar/bat/four.tgz', 'foo/two.tgz', 'one.tgz',
+      'three.txt'], [], [], []))
+
+
+  def test_removed_files_filesystem_snapshot(self):
+
+    after = {}
+
+    snapshot = before_after_filesystem_snapshot.snapshot(self.before, after)
+    self.assertEqual(snapshot, ([], [], [], ['bar/bat/four.tgz', 'foo/two.tgz',
+      'one.tgz', 'three.txt']))
+
+
+  def test_new_filesystem_snapshot(self):
+    after = {
+      'five.tgz': '1234567890defghi',
+      'foo/bar/six.tgz': '0000001111112234',
+      'foofoo/seven.txt': '1111222233334555'
+    }
+
+    snapshot = before_after_filesystem_snapshot.snapshot(self.before, after)
+    self.assertEqual(snapshot, ([], [], ['five.tgz', 'foo/bar/six.tgz',
+      'foofoo/seven.txt'], ['bar/bat/four.tgz', 'foo/two.tgz', 'one.tgz',
+      'three.txt']))
+
+
+  def test_fully_modified_filesystem_snapshot(self):
+
+    after = {
+      'one.tgz': '1234567890aabbcc',
+      'foo/two.tgz': '0000001111112233',
+      'three.txt': '1111222233334455',
+      'bar/bat/four.tgz': '6677889900123456'
+    }
+
+    snapshot = before_after_filesystem_snapshot.snapshot(self.before, after)
+    self.assertEqual(snapshot, ([], ['bar/bat/four.tgz', 'foo/two.tgz',
+      'one.tgz', 'three.txt'], [], []))
+
+
+  def test_partially_modified_filesystem_snapshot(self):
+
+    after = {
+      'five.txt': '5555555555555555',
+      'one.tgz': '1234567890abcdef',
+      'foo/two.tgz': 'ffffffffffffffff',
+      'bar/bat/four.tgz': '6677889900123456',
+      'baz/six.tgz': '6666666666666666'
+    }
+
+    snapshot = before_after_filesystem_snapshot.snapshot(self.before, after)
+    self.assertEqual(snapshot, (['one.tgz'], ['bar/bat/four.tgz',
+      'foo/two.tgz'], ['baz/six.tgz', 'five.txt'], ['three.txt']))
+
+  def test_generate_artifact_rules(self):
+
+    after = {
+      'five.txt': '5555555555555555',
+      'one.tgz': '1234567890abcdef',
+      'foo/two.tgz': 'ffffffffffffffff',
+      'bar/bat/four.tgz': '6677889900112233',
+      'baz/six.tgz': '6666666666666666'
+    }
+
+    artifact_rules = {
+      'expected_materials': [
+        ['ALLOW', 'one.tgz'],
+        ['ALLOW', 'bar/bat/four.tgz'],
+        ['ALLOW', 'foo/two.tgz'],
+        ['ALLOW', 'three.txt'],
+        ['DELETE', 'three.txt'],
+        ['DISALLOW', '*']
+      ],
+      'expected_products': [
+        ['ALLOW', 'one.tgz'],
+        ['ALLOW', 'bar/bat/four.tgz'],
+        ['MODIFY', 'foo/two.tgz'],
+        ['CREATE', 'five.txt'],
+        ['CREATE', 'baz/six.tgz'],
+        ['DISALLOW', '*']
+      ]
+    }
+    snapshot = before_after_filesystem_snapshot.snapshot(self.before, after)
+    rules = before_after_filesystem_snapshot.generate_artifact_rules(snapshot)
+    self.assertEqual(sorted(artifact_rules), sorted(rules))
```

Comment:
I doubt that this is what you want to test. See [my comment about calling sorted on a dict](https://github.com/in-toto/layout-web-tool/pull/46#discussion_r464386252). You probably `want assertDictEqual` instead.

Reply:
Oh sorry, I somehow miss your previous comment, but now I understand why I shouldn't sort the dictionary. Thanks!